File: src/utils/tiles.js
// Utilidades para componer la "vista de cámara" a partir de tiles satelitales.

export const ZOOM_CAMARA = 18;

/** Coordenadas de tile (fraccionarias) para un punto [lng, lat]. */
export function tileXY([lng, lat], z = ZOOM_CAMARA) {
  const n = 2 ** z;
  const x = ((lng + 180) / 360) * n;
  const r = (lat * Math.PI) / 180;
  const y = ((1 - Math.log(Math.tan(r) + 1 / Math.cos(r)) / Math.PI) / 2) * n;
  return { x, y };
}

export const tileURL = (x, y, z = ZOOM_CAMARA) =>
  `https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/${z}/${y}/${x}`;

/** Mosaico 3x3 centrado en `pos`, con el desplazamiento sub-tile en %. */
export function mosaico(pos, z = ZOOM_CAMARA) {
  const { x, y } = tileXY(pos, z);
  const bx = Math.floor(x), by = Math.floor(y);
  const tiles = [];
  for (let dy = -1; dy <= 1; dy++) {
    for (let dx = -1; dx <= 1; dx++) tiles.push([bx + dx, by + dy]);
  }
  return { tiles, offX: (x - bx - 0.5) * 100, offY: (y - by - 0.5) * 100, z };
}

File: appmovil/vigidrone/lib/util/tiles.dart
import 'dart:math' as math;
import 'package:latlong2/latlong.dart';

/// Imágenes satelitales de Esri World Imagery: sin API key ni tarjeta,
/// suficientes para el prototipo y para el mapa de la tablet.
const plantillaSatelite =
    'https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}';

const plantillaEtiquetas =
    'https://server.arcgisonline.com/ArcGIS/rest/services/Reference/World_Boundaries_and_Places/MapServer/tile/{z}/{y}/{x}';

String urlTile(int x, int y, int z) =>
    'https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/$z/$y/$x';

/// Coordenadas de tile (fraccionarias) de un punto.
({double x, double y}) tileXY(LatLng p, int z) {
  final n = math.pow(2, z).toDouble();
  final x = (p.longitude + 180) / 360 * n;
  final r = p.latitude * math.pi / 180;
  final y = (1 - math.log(math.tan(r) + 1 / math.cos(r)) / math.pi) / 2 * n;
  return (x: x, y: y);
}

/// Mosaico 3×3 centrado en el punto, con el desfase sub-tile en fracción.
class Mosaico {
  const Mosaico(this.tiles, this.desfaseX, this.desfaseY, this.zoom);

  final List<({int x, int y})> tiles;
  final double desfaseX;
  final double desfaseY;
  final int zoom;

  factory Mosaico.en(LatLng p, {int zoom = 18}) {
    final t = tileXY(p, zoom);
    final bx = t.x.floor(), by = t.y.floor();
    final lista = <({int x, int y})>[];
    for (var dy = -1; dy <= 1; dy++) {
      for (var dx = -1; dx <= 1; dx++) {
        lista.add((x: bx + dx, y: by + dy));
      }
    }
    return Mosaico(lista, t.x - bx - 0.5, t.y - by - 0.5, zoom);
  }
}
